contracts/src/token/erc20/extensions/erc4626.rs

```diff
@@ -0,0 +1,528 @@
+        address indexed receiver,
+        address indexed owner,
+        uint256 assets,
+        uint256 shares
+    );
+}
+
+sol! {
+    /// Indicates an error where a deposit operation failed
+    /// because the supplied `assets` exceeded the maximum allowed for the `receiver`.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+     error ERC4626ExceededMaxDeposit(address receiver, uint256 assets, uint256 max);
+
+    /// Indicates an error where a mint operation failed
+    /// because the supplied `shares` exceeded the maximum allowed for the `receiver`.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC4626ExceededMaxMint(address receiver, uint256 shares, uint256 max);
+
+    /// Indicates an error where a withdrawal operation failed
+    /// because the supplied `assets` exceeded the maximum allowed for the `owner`
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC4626ExceededMaxWithdraw(address owner, uint256 assets, uint256 max);
+
+    /// Indicates an error where a redemption operation failed
+    /// because the supplied `shares` exceeded the maximum allowed for the `owner`.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC4626ExceededMaxRedeem(address owner, uint256 shares, uint256 max);
+}
+
+/// Error type from [`Erc4626`] contract.
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    /// Indicates an error where a deposit operation failed
+    /// because the supplied `assets` exceeded the maximum allowed for the
+    /// `receiver`.
+    ExceededMaxDeposit(ERC4626ExceededMaxDeposit),
+
+    /// Indicates an error where a mint operation failed
+    /// because the supplied `shares` exceeded the maximum allowed for the
+    /// `receiver`
+    ExceededMaxMint(ERC4626ExceededMaxMint),
+
+    /// Indicates an error where a withdrawal operation failed
+    /// because the supplied `assets` exceeded the maximum allowed for the
+    /// `owner`
+    ExceededMaxWithdraw(ERC4626ExceededMaxWithdraw),
+
+    /// Indicates an error where a redemption operation failed
+    /// because the supplied `shares` exceeded the maximum allowed for the
+    /// `owner`.
+    ExceededMaxRedeem(ERC4626ExceededMaxRedeem),
+
+    /// Error type from [`SafeErc20`] contract [`safe_erc20::Error`].
+    SafeErc20(safe_erc20::Error),
+
+    /// Error type from [`Erc20`] contract [`erc20::Error`].
+    Erc20(erc20::Error),
+}
+
+///  ERC4626 Contract.
+#[storage]
+pub struct Erc4626 {
+    /// The ERC20 token
+    pub _asset: Erc20,
+
+    /// The SafeERC20 token
+    pub _safe_erc20: SafeErc20,
+
+    /// The underlying asset's decimals
+    pub _underlying_decimals: StorageU8,
+}
+
+/// NOTE: Implementation of [`TopLevelStorage`] to be able use `&mut self` when
+/// calling other contracts and not `&mut (impl TopLevelStorage +
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl TopLevelStorage for Erc4626 {}
+
+#[public]
+impl IERC4626 for Erc4626 {
```

Comment:
In Solidity, the `Erc4626` extension inherits ERC-20 logic, and also ERC-20 Metadata logic, and therefore exposes its functions. In Stylus, we usually have to get creative.

Since direct inheritance is not possible, and judging by what the needs of this extension are:

- adds new state
- adds new functionality
- uses ERC-20 functionality without overriding it
- reexports ERC-20 Metadata functions, but overrides its `decimals`

We can organize it in the following manner:

- `Erc4626` state includes all of the below:
    - `_asset: Address` - stored as address for ease of use with `SafeErc20`.
    - `erc20: Erc20` - used to access ERC20 functions _without overriding them_.
    - `erc20_metadata: Erc20Metadata` - used to override `decimals`, and reexport `name` and `symbol`.
- Usage example should state that the contract should inherit both `Erc4626` and `Erc20` to access their functions. There should be no need to inherit `Erc20Metadata`, as it is included in `Erc4626`.

This will require some changes to your implementation.

Reply:
@Ifechukwudaniel I edited my original comment